src/streaming/s3.sink.cpp

```diff
@@ -0,0 +1,182 @@
+#include "macros.hh"
+#include "s3.sink.hh"
+
+#include <miniocpp/client.h>
+
+#ifdef min
+#undef min
+#endif
+
+zarr::S3Sink::S3Sink(std::string_view bucket_name,
+                     std::string_view object_key,
+                     std::shared_ptr<S3ConnectionPool> connection_pool)
+  : bucket_name_{ bucket_name }
+  , object_key_{ object_key }
+  , connection_pool_{ connection_pool }
+{
+    EXPECT(!bucket_name_.empty(), "Bucket name must not be empty");
+    EXPECT(!object_key_.empty(), "Object key must not be empty");
+    EXPECT(connection_pool_, "Null pointer: connection_pool");
+}
+
+zarr::S3Sink::~S3Sink()
+{
+    if (!is_multipart_upload_() && nbytes_buffered_ > 0) {
+        if (!put_object_()) {
+            LOG_ERROR("Failed to upload object: %s", object_key_.c_str());
+        }
+    } else if (is_multipart_upload_()) {
+        if (nbytes_buffered_ > 0 && !flush_part_()) {
+            LOG_ERROR("Failed to upload part %zu of object %s",
+                      parts_.size() + 1,
+                      object_key_.c_str());
+        }
+        if (!finalize_multipart_upload_()) {
+            LOG_ERROR("Failed to finalize multipart upload of object %s",
+                      object_key_.c_str());
+        }
+    }
+}
+
+bool
+zarr::S3Sink::write(size_t offset, const uint8_t* data, size_t bytes_of_data)
```

Comment:
We can't always use multipart upload because there's a minimum size (5 MiB) for all but the last part in a multipart series. Our metadata is much smaller than that. The design is intended to treat S3 objects the same way we treat files on the filesystem, to the extent that that's possible. There are two very important distinctions:

1. We can't seek arbitrarily in an S3 object. This is usually fine, as data is always getting written contiguously.
2. Once an S3 object is closed, we can't reopen it and append to it. So if we were to expose `finalize()` to users, then we need to invalidate a Sink after it's been finalized, i.e., disallow users from writing to it.